enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,332 @@
+- Provide the same Kube-api downtime values ​​as in stack implemented only with static pods for master reboot and master drain cases.
+
+### Non-Goals
+
+- Define a new self-hosted stack for DNS or load balancing
+
+## Proposal
+
+The [Keepalived and infra DNS components](https://github.com/openshift/machine-config-operator/tree/master/manifests/on-prem) on the bootstrap node will keep running through the MCO.
+However, the services on masters and workers nodes will be managed by both MCO and the new operator.
+
+The following functionalities are expected as part of the new operator:
+
+- Introduce a CRD used by the operator to manage the self-hosted stack.
+- Fields in CR spec that control the self-hosted stack configuration.
+- Fields in CR status that expose self-hosted stack status (e.g: which node holds the API-VIP).
+- Run Coredns-MDNS, HAProxy and MDNS-publisher self-hosted stack components as K8S DaemonSets (while Keepalived will continue running as static pod)
+
+## Risks and Mitigations
+
+- Impact on on-prem platforms
+- Upgrades/Downgrades
+- Disaster recovery
+- Cluster shutdown and resume/restart
+
+In order to mitigate the impact on on-prem platforms, in fresh deployments all the self-hosted LB, DNS, and VIP services will be enabled, therefore there is no need to rely on external services during deployment, also the on-prem platforms tests should be updated according to Test Plan section described below.
+
+Additionally, a detailed document in which the supported LB, DNS migration processes will be issued.
+
+## Design Details
+
+### Open Questions [optional]
+
+- Is `cluster-hosted-net-services-operator` an acceptable name?
+- In which namespace the new operator components should be hosted?
+
+### current implmentation
+
+In current implementation, the [networking infra componenets](https://github.com/openshift/installer/blob/master/docs/design/baremetal/networking-infrastructure.md) run as static pods through the MCO.
+
+### Suggested design
+
+To support early clustering requirements Keepalived will continue running as static pods through MCO, additionally, a new dnsmasq service (also deployed through MCO) will run in each node while HAProxy, CoreDNS-MDNS and MDNS-publisher will run as daemonset pods through the new operator.
+
+The following table reflects the suggested changes.
+
+| Componenet Name       | Current implementation | Proposed implementation |
+| --------------------- | ---------------------- | ----------------------- |
+| Resolv.conf prepender | NM dispatcher script   | NM dispatcher script    |
+| Keepalived            | Static pod             | Static pod              |
+| HAProxy               | Static pod             | DaemonSet pod           |
+| CoreDNS-MDNS          | Static pod             | DaemonSet pod           |
+| mDNS-Publisher        | Static pod             | DaemonSet pod           |
+| dnsmasq               |   --                   | Systemd service         |
+
```

Comment:
I miss a discussion of why we move to DaemonSets in the first place. But bad gut feeling is still not gone.

Reply:
@rphillips is there an order between static pods and DaemonSets?

Do we know how long it takes until DaemonSets are started? I assume the scheduler is involved cc @soltysh ?

Is a pod ready only when DNS is available?

Reply:
I agree, I would also like to have all the risks of moving to deamonsets written down here. Like the fact that it all depends on KCM to be up and running in order for it to create pods.

Reply:
> Do we know how long it takes until DaemonSets are started? I assume the scheduler is involved cc @soltysh ?

It requires fully functional both kube-controller-manager and kube-scheduler which in turn talk to kube-apiserver through internal LB.

Reply:
It's important to note that both kube-controller-manager and kube-scheduler depend on internal LB to talk to kube-apiserver. Does this switch affect it in any way? It's critical to ensure that during any rollout/change/upgrade, etc. kcm and ks can talk to any other kas in the cluster. We've tried in the past localhost approach and that didn't work well during rollouts and upgrades.

Reply:
> @yboaron what I'm missing in enhancement is:
> 
> * What are the degraded conditions this new operator will eventually report?
> * What kind of healthchecks the operator will perform to assure the system is working as expected?

That's a good point, I'll update the enhancement.

Reply:
We decided that the new operator will provision (based on configuration) the required pods (e.g: HAProxy) as static pods and not as DaemonSet pods.
I'll update the enhancement

Reply:
@sttts Do you think we can cancel the hold?